### PR TITLE
🎨 pydantic2 migration: fixed unit-tests for dynamic-sidecar

### PR DESCRIPTION
**Pull request by GitHK**
## What do these changes do?

Fixes unit tests for `dynamic-sidecar`.

## Dev-ops checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

### CODE REVIEW

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py [hunk 1]
```diff
@@ -31,6 +41,10 @@ class ResourceTrackingSettings(BaseCustomSettings):
         description="each time the status of the service is propagated",
     )
 
+    _legacy_parsing_dynamic_scheduler_stop_service_timeout = (
+        validate_timedelta_in_legacy_mode("RESOURCE_TRACKING_HEARTBEAT_INTERVAL")
```
Comment (sanderegg) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py:45]: did you check this: https://docs.pydantic.dev/latest/api/standard_library_types/#datetimetimedelta
the documentation states that it still works correctly.

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py [hunk 2]
```diff
@@ -100,7 +117,7 @@ class ApplicationSettings(BaseCustomSettings, MixinLoggingSettings):
     )
 
     DYNAMIC_SIDECAR_RESERVED_SPACE_SIZE: ByteSize = Field(
-        parse_obj_as(ByteSize, "10Mib"),
+        TypeAdapter(ByteSize).validate_python("10Mib"),
```
Comment (sanderegg) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py:120]: You created a ByteSizeAdapter, why don't you use it?

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/models/shared_store.py [hunk 3]
```diff
@@ -74,6 +85,7 @@ async def _setup_initial_volume_states(self) -> None:
                 (VolumeCategory.OUTPUTS, VolumeStatus.CONTENT_NEEDS_TO_BE_SAVED),
                 (VolumeCategory.STATES, VolumeStatus.CONTENT_NEEDS_TO_BE_SAVED),
             ]:
+                # pylint: disable=unsupported-assignment-operation
```
Comment (sanderegg) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/models/shared_store.py:88]: why is this necessary? this looks fishy

**Review by sanderegg — COMMENTED** [review]
please check that stuff with time delta

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py (hunk 1 above)
Comment (GitHK) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py:45]: The issue is that we were providing str timedeltas in env vars with a different format (the number of seconds). Now this is no longer accepted, but it used to work with pydantic1

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/models/shared_store.py (hunk 3 above)
Comment (GitHK) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/models/shared_store.py:88]: no clue, had to use it in a couple of places @giancarloromeo any idea why?

File: packages/common-library/src/common_library/pydantic_settings_validators.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import datetime
+
+from pydantic import field_validator
+
+
+def _get_float_string_as_seconds(
```
Comment (pcrespov) [inline packages/common-library/src/common_library/pydantic_settings_validators.py:6]: The function name is really unclear, and it doesn't clearly convey the intended behavior. Based on the code, the input-output mapping appears to work as follows:

- For `string` inputs: it attempts to convert to `float`; if conversion fails, it returns the original `string`.
- For `float` inputs: it returns the `float` unchanged (bypass).
- For `timedelta` inputs: it returns the `timedelta` unchanged (bypass).

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py (hunk 2 above)
Comment (GitHK) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py]: I've replaced it globally

File: packages/models-library/src/models_library/services_creation.py [hunk 5]
```diff
@@ -36,10 +41,18 @@ class CreateServiceMetricsAdditionalParams(BaseModel):
...
+                "service_version": ServiceVersionTypeAdapter.validate_python("0.0.1"),
                 "service_resources": {},
                 "service_additional_metadata": {},
+                "pricing_unit_cost_id": None,
             }
         }
     )
+
+
+CreateServiceMetricsAdditionalParamsTypeAdapter = TypeAdapter(
```
Comment (pcrespov) [inline packages/models-library/src/models_library/services_creation.py:56]: THOUGHT: i find more readable the composed name instead of ending up having many huge alias names. 
Similar situation as with `Page[Model]` rather than an alias `PageModel`

File: packages/models-library/src/models_library/services_io.py [hunk 6]
```diff
@@ -259,3 +258,6 @@ def from_json_schema(cls, port_schema: dict[str, Any]) -> "ServiceOutput":
         """Creates output port model from a json-schema"""
         data = cls._from_json_schema_base_implementation(port_schema)
         return cls.model_validate(data)
+
+
+ServiceOutputTypeAdapter = TypeAdapter(ServiceOutput)
```
Comment (pcrespov) [inline packages/models-library/src/models_library/services_io.py:263]: same here

**Review by pcrespov — APPROVED** [review]

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/models/shared_store.py (hunk 3 above)
Comment (giancarloromeo) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/models/shared_store.py:88]: Static analysis tools struggle with dynamic frameworks like Pydantic. It can be safely disabled, or if you don't want to disable it, ensure that the `self.volume_states` is a `dict` (i.e. with `__setitem__` method) adding a check. But this will result more pedantic than pydantic.

```python
    volume_states: dict[VolumeCategory, VolumeState] = Field(
        default_factory=dict, description="persist the state of each volume"
    )

    async def _setup_initial_volume_states(self) -> None:
        async with self:
            for category, status in [
                (VolumeCategory.INPUTS, VolumeStatus.CONTENT_NO_SAVE_REQUIRED),
                (VolumeCategory.SHARED_STORE, VolumeStatus.CONTENT_NO_SAVE_REQUIRED),
                (VolumeCategory.OUTPUTS, VolumeStatus.CONTENT_NEEDS_TO_BE_SAVED),
                (VolumeCategory.STATES, VolumeStatus.CONTENT_NEEDS_TO_BE_SAVED),
            ]:
                if self.volume_states is None:
                    self.volume_states = {}
                self.volume_states[category] = VolumeState(status=status)

```

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py (hunk 1 above)
Comment (sanderegg) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py:39]: so the naming is really weird. I think we do not care that it might be legacy or so. this should become a real validator converting a numeric string into seconds.

**Review by giancarloromeo — APPROVED** [review]
Just few comments. Thank you very much for your support here.

**Comment by giancarloromeo** [comment]
There are still failing tests. And a bunch of deprecated `parse_obj` and `parse_obj_as` that is better to migrate too.

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py (hunk 1 above)
Comment (sanderegg) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py:39]: alright, so what I did for the autoscaling, is go away from that float/int nonsense.
I changed them to follow ISO 8601: "00:00:02" if you want 2 seconds. and that is it

File: packages/common-library/src/common_library/pydantic_type_adapters.py [hunk 7]
```diff
@@ -1,8 +1,12 @@
 from typing import Final
 
 from common_library.pydantic_networks_extension import AnyHttpUrlLegacy, AnyUrlLegacy
-from pydantic import TypeAdapter
+from pydantic import ByteSize, TypeAdapter
 
 AnyUrlLegacyAdapter: Final[TypeAdapter[AnyUrlLegacy]] = TypeAdapter(AnyUrlLegacy)
 
-AnyHttpUrlLegacyAdapter: Final[TypeAdapter[AnyHttpUrlLegacy]] = TypeAdapter(AnyHttpUrlLegacy)
+AnyHttpUrlLegacyAdapter: Final[TypeAdapter[AnyHttpUrlLegacy]] = TypeAdapter(
```
[the remark this answers is not in the document]
Comment (GitHK) [reply packages/common-library/src/common_library/pydantic_type_adapters.py:8]: for now I'll stick to Fina and *Adapter

File: packages/models-library/src/models_library/services_creation.py (hunk 5 above)
Comment (GitHK) [reply packages/models-library/src/models_library/services_creation.py]: I have the same opinion as you but people pointed out the "performance" hit in instantiating the objects. So we came up with this alternative. We have to solutions:
1. stick with `TypeAdapter(Model)`
2. go with this proposal for optimisation reason

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py (hunk 1 above)
Comment (GitHK) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py]: we also have to convert all the settings int env vars repos to ensure this is still working. And I did not want to mess up also with those, that's the reason for this validator.
let me rename it a bit and see if it's better, at least

Comment (sanderegg) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py]: well it's a migration.. and there are many of these interval where the validator was not setup.

File: packages/common-library/tests/test_pydantic_settings_validators.py
Comment (sanderegg) [inline packages/common-library/tests/test_pydantic_settings_validators.py:1]: this file should be renamed right?

File: packages/models-library/src/models_library/api_schemas_dynamic_sidecar/containers.py [hunk 8]
```diff
@@ -15,4 +15,7 @@ class ActivityInfo(BaseModel):
     )
 
 
+ActivityInfoAdapter: Final[TypeAdapter[ActivityInfo]] = TypeAdapter(ActivityInfo)
```
Comment (sanderegg) [inline packages/models-library/src/models_library/api_schemas_dynamic_sidecar/containers.py:18]: Is there not something akin to C++ templates for this?

File: packages/pytest-simcore/src/pytest_simcore/helpers/playwright_sim4life.py [hunk 9]
```diff
@@ -5,9 +5,8 @@
 from typing import Final, TypedDict
 
 import arrow
+from common_library.pydantic_type_adapters import ByteSizeAdapter
```
Comment (sanderegg) [inline packages/pytest-simcore/src/pytest_simcore/helpers/playwright_sim4life.py:8]: be careful here. I'm not sure you are not adding that dependency in the playwright stuff with this..

File: scripts/maintenance/computational-clusters/autoscaled_monitor/core.py [hunk 10]
```diff
@@ -10,9 +10,10 @@
 import parse
 import rich
 import typer
+from common_library.pydantic_type_adapters import ByteSizeAdapter
```
Comment (sanderegg) [inline scripts/maintenance/computational-clusters/autoscaled_monitor/core.py:13]: can you please revert all the changes in here?
This is a standalone package that does not need nore want to depend on anything.

**Review by sanderegg — CHANGES_REQUESTED** [review]
Please revert the changes in scripts/maintenance/computational-clusters. Thanks!

File: scripts/maintenance/computational-clusters/autoscaled_monitor/core.py (hunk 10 above)
Comment (GitHK) [reply scripts/maintenance/computational-clusters/autoscaled_monitor/core.py:13]: ohh boy missed this one. thanks

File: packages/models-library/src/models_library/api_schemas_dynamic_sidecar/containers.py (hunk 8 above)
Comment (GitHK) [reply packages/models-library/src/models_library/api_schemas_dynamic_sidecar/containers.py:18]: Would you mind clarifying this? Could I have an example of what you are searching for?

File: packages/pytest-simcore/src/pytest_simcore/helpers/playwright_sim4life.py (hunk 9 above)
Comment (GitHK) [reply packages/pytest-simcore/src/pytest_simcore/helpers/playwright_sim4life.py:8]: let's revert this one

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py (hunk 1 above)
Comment (GitHK) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py]: applied everywhere and added a nice warning message
```
tests/unit/test_core_settings.py::test_settings_with_envdevel_file
  /home/silenthk/work/pr-osparc-pydantic-v2-dynamic-sidecar/services/dynamic-sidecar/tests/unit/test_core_settings.py:15: DeprecationWarning: RESOURCE_TRACKING_HEARTBEAT_INTERVAL='33444' --should be set to--> RESOURCE_TRACKING_HEARTBEAT_INTERVAL='PT9H17M24S' (ISO8601 datetime format). Please also convert the value in the >>OPS REPOSITORY<<. For details: https://docs.pydantic.dev/1.10/usage/types/#datetime-types.
    assert ApplicationSettings.create_from_envs()
```

**Review by sanderegg — APPROVED** [review]
thanks for all the changes
